[css-anchor-position-1] Implement scroll position adjustments

anttijk opened the pull request:
#### 69b703409e424497c2dfc96feed959984a9048bc
<pre>
[css-anchor-position-1] Implement scroll position adjustments
<a href="https://bugs.webkit.org/show_bug.cgi?id=281139">https://bugs.webkit.org/show_bug.cgi?id=281139</a>
<a href="https://rdar.apple.com/137590987">rdar://137590987</a>

Reviewed by Antoine Quint.

<a href="https://drafts.csswg.org/css-anchor-position-1/#scroll">https://drafts.csswg.org/css-anchor-position-1/#scroll</a>

Anchor positioned element may be on a different scroller than the anchor. In this case
adjustments are needed to keep the positions in sync.

This patch does not yet implement scrolling adjustments in the scrolling tree so
the exact anchoring can&apos;t be maintained in all cases during asynchronous scrolling.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-cleanup-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-cleanup.html: Added.

Add a test for cleanups.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-011-expected.txt:

Note that @position-try is not implemented so neither fails nor passes are meaningful.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):

Update the snapshotted scroll offsets on rendering update if needed.

(WebCore::operator&lt;&lt;):
* Source/WebCore/page/Page.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateTransformFromStyle const):

Includes snapshotted scroll offset to the layer transform.
It is specified to be treated like another transform.

(WebCore::RenderLayer::updateTransform):
(WebCore::RenderLayer::setSnapshottedScrollOffsetForAnchorPositioning):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateTransform):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingLayer const):
(WebCore::RenderLayerCompositor::requiresOwnBackingStore const):
(WebCore::RenderLayerCompositor::reasonsForCompositing const):
(WebCore::compositingReasonToString):
(WebCore::RenderLayerCompositor::requiresCompositingForAnchorPositioning const):

Composite anchor positioned elements that require adjustments.

* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::scrollTo):

Update the snapshotted scroll offsets during scrolling if needed.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::shouldUseTransformFromContainer const):
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::offsetFromAncestorContainer):

Substract away the effect so offset is as if all scrollers were on (0, 0) scroll position.

(WebCore::Style::scrollOffsetFromAncestorContainer):
(WebCore::Style::AnchorPositionEvaluator::updateSnapshottedScrollOffsets):
* Source/WebCore/style/AnchorPositionEvaluator.h:

* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::clearAnchorPositioningState):

Reset the offsets when clearing the state.
Currently we build these structures from scratch if anything changes.

Canonical link: <a href="https://commits.webkit.org/285394@main">https://commits.webkit.org/285394@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc4829a8e9ff412d7c817dabf706dba4d175d2a9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72531 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51952 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25319 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76715 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23740 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74646 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59756 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23562 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15618 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75598 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47021 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62463 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/37531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43672 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/19932 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22090 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/65540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20288 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78384 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16772 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/19416 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16820 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62476 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16009 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13102 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6750 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47750 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/2535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48819 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50109 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48562 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/Page.h:
```diff
@@ -280,6 +280,7 @@ enum class RenderingUpdateStep : uint32_t {
 #endif
     RestoreScrollPositionAndViewState = 1 << 24,
     AdjustVisibility                  = 1 << 25,
+    SnapshottedScrollOffsets          = 1 << 26,
```
I think it would be nice to keep these in the order in which `Page::updateRendering()` consults them.

anttijk replied:
Existing ones are not in order either. I moved the offset update last though.

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -1330,7 +1333,9 @@ class RenderLayer final : public CanMakeSingleThreadWeakPtr<RenderLayer>, public
     std::unique_ptr<TransformationMatrix> m_transform;
-    
+
+    std::optional<LayoutSize> m_snapshottedScrollOffsetForAnchorPositioning;
```
I don't see this ever getting cleared, when a layer stops doing anchor positioning. Can you add a test for this?

smfr replied:
Also, what happens when a scroller appears or disappears between the anchored and anchoring elements?

anttijk replied:
Good point. Added a test and made sure the the offset is cleared.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -4221,6 +4227,16 @@ bool RenderLayer::participatesInPreserve3D() const
+
+    // FIXME: Scroll offset should be adjusted in the scrolling tree so layers stay exactly in sync.
+    m_snapshottedScrollOffsetForAnchorPositioning = offset;
+    updateTransform();
```
Do we need to eagerly call `updateTransform()` here? We already call it twice per layout (once from style change, once from post-layout).

anttijk replied:
Yes, there is nothing that guarantees it is updated by someone else.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -3352,6 +3357,7 @@ static ASCIILiteral compositingReasonToString(CompositingReason reason)
     case CompositingReason::Root: return "root"_s;
     case CompositingReason::Model: return "model"_s;
     case CompositingReason::BackdropRoot: return "backdrop root"_s;
+    case CompositingReason::AnchorPositioning: return "anchor positioning"_s;
```
I think these reasons propagate out into web inspector code (they show in the Layers tab) so I suspect there is more to do here.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -3940,6 +3946,11 @@ bool RenderLayerCompositor::requiresCompositingForOverflowScrolling(const Render
 
+bool RenderLayerCompositor::requiresCompositingForAnchorPositioning(const RenderLayer& layer) const
+{
+    return !!layer.snapshottedScrollOffsetForAnchorPositioning();
```
`has_value()`?

anttijk replied:
why?

graouts replied:
Is the `!!` even necessary here? From https://en.cppreference.com/w/cpp/utility/optional:
> When an object of type optional<T> is [contextually converted to bool](https://en.cppreference.com/w/cpp/language/implicit_conversion#Contextual_conversions), the conversion returns true if the object contains a value and false if it does not contain a value.

anttijk replied:
not really but I like to make boolean conversions explicit so they don't look like accidents

anttijk replied:
Actually it is necessary, implicit conversion does not happen in this context.

rniwa commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -543,4 +565,47 @@ void AnchorPositionEvaluator::findAnchorsForAnchorPositionedElement(Ref<const El
+
+    auto& states = document.styleScope().anchorPositionedStates();
+    for (auto elementAndState : states) {
+        auto& anchorPositionedElement = elementAndState.key;
```
Can we use Ref here instead?
https://ews-build.s3-us-west-2.amazonaws.com/macOS-Safer-CPP-Checks-EWS/e09966ca-846/scan-build-output/StaticAnalyzerRegressions/WebCore/StaticAnalyzerReports/report-AnchorPositionEvaluator.cpp-updateSnapshottedScrollOffsets-5-408e86.html#EndPath

smfr commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -226,6 +234,20 @@ static LayoutSize offsetFromAncestorContainer(const RenderElement& descendantCon
+    auto offset = LayoutSize { };
+    for (auto* ancestor = descendant.container(); ancestor; ancestor = ancestor->container()) {
+        if (auto* box = dynamicDowncast<RenderBox>(ancestor))
+            offset -= toLayoutSize(box->scrollPosition());
```
This is probably wrong with RTL scrollers.

anttijk replied:
I don't think so, at least not for any obvious reasons. This is the same adjustment done in RenderObject::offsetFromContainer (we reverse it in offsetFromAncestorContainer() and re-apply it in scrollOffsetFromAncestorContainer). 

There are certainly RTL issues with the current code (we fails some existing tests) but those are probably better looked separately.

graouts commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -586,4 +608,55 @@ void AnchorPositionEvaluator::updateAnchorPositioningStatesAfterInterleavedLayou
+
+void AnchorPositionEvaluator::cleanupAnchorPositionedState(Element& element)
+{
+    if (element.document().styleScope().anchorPositionedStates().remove(element)) {
```
I don't know whether it's relevant here, but maybe use `protectedDocument()`?

graouts commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -215,6 +216,13 @@ static LayoutSize offsetFromAncestorContainer(const RenderElement& descendantCon
+        // https://drafts.csswg.org/css-anchor-position-1/#scroll
+        // "anchor() is defined to assume all the scroll containers between the anchor element and
+        // the positioned element’s containing block are at their initial scroll position,"
+        if (auto* boxContainer = dynamicDowncast<RenderBox>(*nextContainer))
```
`CheckedPtr`?

anttijk committed:
[css-anchor-position-1] Implement scroll position adjustments
https://bugs.webkit.org/show_bug.cgi?id=281139
rdar://137590987

Reviewed by Antoine Quint.

https://drafts.csswg.org/css-anchor-position-1/#scroll

Anchor positioned element may be on a different scroller than the anchor. In this case
adjustments are needed to keep the positions in sync.

This patch does not yet implement scrolling adjustments in the scrolling tree so
the exact anchoring can't be maintained in all cases during asynchronous scrolling.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-cleanup-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-cleanup.html: Added.

Add a test for cleanups.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-011-expected.txt:

Note that @position-try is not implemented so neither fails nor passes are meaningful.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):

Update the snapshotted scroll offsets on rendering update if needed.

(WebCore::operator<<):
* Source/WebCore/page/Page.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateTransformFromStyle const):

Includes snapshotted scroll offset to the layer transform.
It is specified to be treated like another transform.

(WebCore::RenderLayer::updateTransform):
(WebCore::RenderLayer::setSnapshottedScrollOffsetForAnchorPositioning):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateTransform):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingLayer const):
(WebCore::RenderLayerCompositor::requiresOwnBackingStore const):
(WebCore::RenderLayerCompositor::reasonsForCompositing const):
(WebCore::compositingReasonToString):
(WebCore::RenderLayerCompositor::requiresCompositingForAnchorPositioning const):

Composite anchor positioned elements that require adjustments.

* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::scrollTo):

Update the snapshotted scroll offsets during scrolling if needed.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::shouldUseTransformFromContainer const):
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::offsetFromAncestorContainer):

Substract away the effect so offset is as if all scrollers were on (0, 0) scroll position.

(WebCore::Style::scrollOffsetFromAncestorContainer):
(WebCore::Style::AnchorPositionEvaluator::updateSnapshottedScrollOffsets):
* Source/WebCore/style/AnchorPositionEvaluator.h:

* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::clearAnchorPositioningState):

Reset the offsets when clearing the state.
Currently we build these structures from scratch if anything changes.

Canonical link: https://commits.webkit.org/285394@main
